### PR TITLE
[Sketcher] Stage 5 of refactors

### PR DESCRIPTION
**Pull request by AjinkyaDahale**
Draft changes for some refactoring as part of the grant. Now writing some tests.

## Issues

N/A

## Before and After Images

N/A

### CODE REVIEW

**Comment by AjinkyaDahale** [comment]
I'll mark this PR as ready for review---despite some items not having the tests I'd like to add---since they are mostly still just rewrites of conditionals and loops, as well as some extractions into functions.

At a minimum, the commits affecting `replaceGeometries()` must be merged into 1.1.0, since they fix a major segfault. This happens only when the method is used in functions like `join` that replace more geometries with fewer, which so far do not use the method.

**Comment by kadet1090** [comment]
@PaddleStroke can you take a look at that one? I'll go through it in terms of code quality issues but my knowledge about sketcher is not enough to spot business logic issues.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 1]
```diff
@@ -4190,143 +4163,146 @@ int SketchObject::addSymmetric(const std::vector<int>& geoIdList, int refGeoId,
...
-                    }
+            };
+
+            if (constr->Third == GeoEnum::GeoUndef) {
+                if (!(constr->Type == Sketcher::Coincident
+                      || constr->Type == Sketcher::Perpendicular
+                      || constr->Type == Sketcher::Parallel || constr->Type == Sketcher::Tangent
+                      || constr->Type == Sketcher::Distance || constr->Type == Sketcher::Equal
+                      || constr->Type == Sketcher::Angle || constr->Type == Sketcher::PointOnObject
+                      || constr->Type == Sketcher::InternalAlignment)) {
```
Comment (kadet1090) [inline src/Mod/Sketcher/App/SketchObject.cpp]: Can we put each condition in separate line? I know that this is how clang-format does (damn you clang-format!) but it is really hard to read. One trick you can use is to put empty comment at the end of each line to force line break.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 2]
```diff
@@ -7221,112 +7216,96 @@ int SketchObject::delExternal(const std::vector<int>& ExtGeoIds)
...
+        std::unique_ptr<Constraint> newCstr(cstr->clone());
+        for (auto GeoId : geoIds) {
+            GeoId += offset++;
+            if (newCstr->First >= GeoId && newCstr->Second >= GeoId && newCstr->Third >= GeoId) {
+                break;
             }
-            newConstraints.push_back(cstr);
+            changeConstraintAfterDeletingGeo(newCstr.get(), GeoId);
         }
+        newConstraints.push_back(cstr);
```
Comment (PaddleStroke) [inline src/Mod/Sketcher/App/SketchObject.cpp]: Bug here. Before we had `cstr = cstr->clone();` so `newConstraints.push_back(cstr);` was correct.
Now you made `std::unique_ptr<Constraint> newCstr(cstr->clone());` yet your are still `newConstraints.push_back(cstr);`

**Review by PaddleStroke — CHANGES_REQUESTED** [review]
Considering the regressions caused by the previous refactoring PR and the critical bug I found, I think this should wait for after 1.1 release since the chances to introduce regressions exist and it does not fix any bug so there is no rush getting it in.

Aside from that I notice significant behavior changes in join() and replaceGeometries() but I guess those are voluntary changes.

Another thing is about `addCopy()` `addSymmetric()` `removeAxesAlignment()` and `transferFilletConstraints()`. These sections involve complete logical rewrites. While they appear to be functionally equivalent and are much cleaner, the risk of introducing subtle bugs is high. They require extensive testing. 

Given the regressions caused by the previous PRs, the massive size of this PR and the ease with which this could be split up, I do not see how this could be reasonably be merged as one block. Because as one block the reviewer ability to review is spread accross the whole thing, and so it is very likely that many things are going to go unnoticed. For example Kadet did not detect the critical bug I detected, and I probably detected it by luck and there are probably more. This PR should be split, one PR for each function so that it can be reviewed properly.

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 2 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp]: So now it should be as follows?

```suggestion
        newConstraints.push_back(newCstr);

```

Comment (PaddleStroke) [reply src/Mod/Sketcher/App/SketchObject.cpp]: I believe that yes or newCstr is useless no? But ultimately you did the refactoring so you should be better placed to confirm for sure.

Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Correct then. Use of suggestion feature would have made it clearer. I probably mixed up with another concept where `PropertyConstraintList::setValues()` (or whatever the class is called) checks that the constraint object already is in the list so does not delete it. I also wonder if that could be used to operate directly on the constraints rather than cloning them, for efficiency gains. That's for another PR.

**Comment by AjinkyaDahale** [comment]
> Considering the regressions caused by the previous refactoring PR and the critical bug I found, I think this should wait for after 1.1 release since the chances to introduce regressions exist and it does not fix any bug so there is no rush getting it in.

Fair enough, I'll leave it to the merge team.

> ...I notice significant behavior changes in join() and replaceGeometries() but I guess those are voluntary changes.

Could you elaborate on what you mean by "voluntary changes"? The change in `replaceGeometries()` should be added sooner than later.
 
> This PR should be split, one PR for each function so that it can be reviewed properly.

This would be quite an overhead, but doable if the reviewer(s) want it that way.

**Comment by PaddleStroke** [comment]
> Could you elaborate on what you mean by "voluntary changes"? The change in replaceGeometries() should be added sooner than later.

I don't say those changes should not be done. I just mentioned that there it's not a 1 to 1 refactoring, ie there are some logic changes. And I said voluntary to mean that you did them on purpose.

**Comment by kadet1090** [comment]
Refactors can always bring regressions and while they are massively needed it will be probably better to wait after 1.2 to reduce risks.

File: src/Mod/Sketcher/App/SketchObject.cpp [hunk 3]
```diff
@@ -2516,12 +2514,13 @@ void SketchObject::transferFilletConstraints(int geoId1, PointPos posId1, int ge
...
-                bool line1 = c->First == geoId1 && c->FirstPos == PointPos::none;
-                bool line2 = c->First == geoId2 && c->FirstPos == PointPos::none;
-                if (line1 || line2) {
-                    deleteme.push_back(i);
-                }
+            if (c->Type != Sketcher::Distance && c->Type != Sketcher::Equal) {
+                continue;
+            }
+            bool line1 = c->First == geoId1 && c->FirstPos == PointPos::none;
+            bool line2 = c->First == geoId2 && c->FirstPos == PointPos::none;
```
Comment (PaddleStroke) [inline src/Mod/Sketcher/App/SketchObject.cpp:2593]: You should use the new getGeoId/setGeoId etc everywhere you refactor the code. First and FirstPos etc are deprecated.

Comment (FlachyJoe) [reply src/Mod/Sketcher/App/SketchObject.cpp:2593]: Can't we add a deprecated warning in the code?

**Comment by maxwxyz** [comment]
@AjinkyaDahale could you take a look at the CI failure?

**Comment by AjinkyaDahale** [comment]
@maxwxyz from what I can tell it is an issue in FemGui, which I don't touch. It also doesn't happen anywhere else. Maybe a rerun will fix it?

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 2 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Thought I had made the changes. Done just now.

**Comment by AjinkyaDahale** [comment]
Can this be merged into `main` now, since 1.1 branch is separate? If I haven't already  fixed the issue mentioned by @PaddleStroke, I can do so shortly.

**Comment by maxwxyz** [comment]
@AjinkyaDahale the PR has a conflict

**Comment by AjinkyaDahale** [comment]
@maxwxyz I can resolve them once you're ready to merge. Doing that now could mean I have to dot that again.

**Comment by maxwxyz** [comment]
@chennes are your ready?

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 1 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp]: Done

File: src/Mod/Sketcher/App/SketchObject.cpp (hunk 3 above)
Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp:2593]: As much as I'd like this to be the case, this will take longer, and I'd rather not have to constantly rebase. Any automation possible?

Comment (AjinkyaDahale) [reply src/Mod/Sketcher/App/SketchObject.cpp:2593]: @hyarion can you address adding the deprecated warning?

**Comment by PaddleStroke** [comment]
This PR broke the symmetry tool. See https://github.com/FreeCAD/FreeCAD/pull/26886 (Thanks @AjinkyaDahale for fixing it promptly).

I'd like to take the opportunity to say that [I was right](https://github.com/FreeCAD/FreeCAD/pull/22951#pullrequestreview-3195292513) and you guys (@FreeCAD/maintainers) didn't listen to me (again). This PR was too big to be merged as one entity and should have been split so that it could be reviewed/tested correctly.

The result? It has not been tested at all. And so it was very likely to generate obvious regressions such as that one where the symmetric tool is completely broken. This is a very obvious regression, if the refactor of addSymmetric was a standalone PR, it would have been detected before merge.

The regression is not problem per say, that happens. The problem is that you ignored what was a good review asking something totally reasonable that would have prevented it.

Unrelated to this but I take the opportunity to state it, I asked 3 times for https://github.com/FreeCAD/FreeCAD/pull/25513 to be merged ASAP in 1.1 and got ignored yet again. So don't be surprise when the situation is a mess when that PR merge in 1.2.

**Comment by AjinkyaDahale** [comment]
> This PR broke the symmetry tool. See #26886 (Thanks @AjinkyaDahale for fixing it promptly).

Glad this was found out this soon.

> I'd like to take the opportunity to say that [I was right](https://github.com/FreeCAD/FreeCAD/pull/22951#pullrequestreview-3195292513) and you guys (@FreeCAD/maintainers) didn't listen to me (again).

I would say that warning was indeed heeded, which is why this PR was merged for 1.2, not 1.1. 

> [from that review] While [some modified methods specified earlier] appear to be functionally equivalent and are much cleaner, the risk of introducing subtle bugs is high. They require extensive testing.

Despite care (and in many cases, explicit tests) from my end, there are bugs that can be introduced. Unfortunately, at some point, I need to put the onus of the "extensive testing" on the bleeding edge user base. They have called it quite often for `trim`, which has been a much bigger rewrite than other methods, and the bugs have been found.

> This PR was too big to be merged as one entity and should have been split so that it could be reviewed/tested correctly.

> The result? It has not been tested at all. And so it was very likely to generate obvious regressions such as that one where the symmetric tool is completely broken. This is a very obvious regression, if the refactor of addSymmetric was a standalone PR, it would have been detected before merge.

I doubt there was much practicality in splitting the PR into such fine detail. At some point, I would have preferred that the reviewers look at this commit by commit, treating each as a separate PR (which is what the situation would be anyway if the split were done). This might not apply to other PRs, but here I believe I made a reasonable effort that each commit was touching a separate part of the code, and one part of the code wasn't unnecessarily changed multiple times.

The biggest reason is that many changes were made fairly organically before being organized on a per-method basis. I did try to make a reasonable effort post-facto to keep each commit separate, but that would lead to a lot of conflicts and possible re-reading and re-writing, which is why I left things as is. This particular `addSymmetric` change, I tried to revert just the one commit that explicitly did the refactor, but the regression didn't go.

Even if the above was not the case, changes in individual methods sometimes need changes in other methods that it calls. It is highly possible that multiple method refactors need different (and possibly conflicting) changes in the same "upstream" method. If these are merged out of order, there may be conflicts and regressions. If they are in order, it is the same as a single PR.

> The regression is not problem per say, that happens. The problem is that you ignored what was a good review asking something totally reasonable that would have prevented it.

Again, I believe you were not ignored here, and I have stated my opinion on keeping this a single PR. I will make reasonable efforts to make future refactor PRs smaller, but no guarantees. If there are other aspects that you believe were ignored, please specify.

> Unrelated to this but I take the opportunity to state it, I asked 3 times for #25513 to be merged ASAP in 1.1 and got ignored yet again. So don't be surprise when the situation is a mess when that PR merge in 1.2.

I haven't been involved with this, but there is a similarity with #25513 and this PR, in that waiting to merge this one may have caused a mess already. It's a coin toss whether the issue in `addSymmetric` came because of my original commit, or because of a botched conflict resolution. Further rebases may have added more issues, which is why I was insisting on limiting them (of course, in addition to being tired of doing it so often).

**Comment by PaddleStroke** [comment]
> Again, I believe you were not ignored here, and I have stated my opinion on keeping this a single PR.

By 'you' (in "that you ignored") I didn't you Ajinkya, I meant the maintainers. I mentioned this should be split. You argued that it's difficult but up to the maintainers. They merged. So really I don't think you are at fault.

Yes the PR is nicely split, each commit on a different subject. But the fact remain that this is 23 commits. It was sure that they would not review / test each commit individually. Hence my suggestion this should be split. Because as I stated, a PR focused on 'addSymmetry' would have received at least a minimal test of the tool and so finding out the obvious regression.

Again I'm not throwing a stone because a regression was made. Inacuracies are bound to happen. Especially as you stated the rebase make it more tiresome and difficult.

It's just that since I was asked to review, gave my opinion and ultimately it was (admitedly partly since it waited for 1.2) overlooked, I take the opportunity to point out that I was right and should have been listen to.

**Comment by PaddleStroke** [comment]
[Another crash](https://github.com/FreeCAD/FreeCAD/pull/27693) caused by this PR. This is the 4th regression I have to deal with because of that PR. 

In this one it's a missing early return and failure to handle array shrinkage in delExternalPrivate.

@AjinkyaDahale please feed one by one the commits of this PR to aistudio.google.com and ask it to check that each of those commits did not break something and do not include any typo. Gemini 3 is very good at spotting such innacuracy or problem.
Just do https://github.com/FreeCAD/FreeCAD/commit/78cdd8d59d42da91c44df559037081800a68a819.diff copy the raw diff of the commit, paste it to gemini and ask 'please check this refactor for any error'.

It needs to be done one commit at a time so that it can focus on one commit specifically.

**Comment by AjinkyaDahale** [comment]
> [Another crash](https://github.com/FreeCAD/FreeCAD/pull/27693) apparantly caused by this PR. A missing early return in delExternalPrivate.

From a cursory look, that PR does not seem to affect any files changed in this PR. Could you elaborate?

> @AjinkyaDahale can you please feed one by one the commits of this PR to aistudio.google.com...

Will take a look. Does it need the context of the whole source code or just these changes are enough?
